File: apps/admin/src/data/articles.ts
import { Article } from "../types/article";

export const initialArticles: Article[] = [
  {
    id: 1,
    title: "从零实现一个兼顾自动发布与视觉表达的个人博客",
    category: "项目实战",
    status: "已发布",
    updated: "今天 09:42",
    excerpt: "记录架构、界面和自动发布流程。",
    content:
      "一次线上故障之后，真正值得留下来的不只是修复结果，还有定位路径、判断依据和可以复用的工程经验。",
  },
  {
    id: 2,
    title: "用 CSS 网格做一个不无聊的文章列表",
    category: "前端开发",
    status: "草稿",
    updated: "昨天 18:20",
    excerpt: "关于信息密度、留白和视觉节奏的一次小实验。",
    content: "好的界面应该帮助读者建立节奏。动效、留白和层级都需要围绕内容本身工作。",
  },
  {
    id: 3,
    title: "技术分享不只是把答案贴出来",
    category: "思考记录",
    status: "已发布",
    updated: "2026-08-26",
    excerpt: "把背景、过程和失败也写进去。",
    content: "还没有答案的问题也值得记录。它们会成为下一次学习和验证的入口。",
  },
];

File: services/api/app/routers/articles.py
from fastapi import APIRouter, HTTPException, Query

from ..models import Article, ArticleCreate, ArticleStatusUpdate, ArticleUpdate, Status
from ..store import (
    create_article,
    delete_article,
    fetch_article,
    fetch_article_by_slug,
    list_articles,
    set_article_status,
    update_article,
)

router = APIRouter(prefix="/api/articles", tags=["articles"])


@router.get("", response_model=list[Article])
def get_articles(
    status: Status | None = Query(default=None),
    q: str | None = Query(default=None, description="Search by title, slug, category, excerpt, or content"),
    limit: int = Query(default=50, ge=1, le=100),
    offset: int = Query(default=0, ge=0),
) -> list[Article]:
    return list_articles(status=status, query=q, limit=limit, offset=offset)


@router.get("/{article_id}", response_model=Article)
def get_article(article_id: int) -> Article:
    try:
        return fetch_article(article_id)
    except LookupError as error:
        raise HTTPException(status_code=404, detail="Article not found") from error


@router.get("/slug/{slug}", response_model=Article)
def get_article_by_slug(slug: str) -> Article:
    try:
        return fetch_article_by_slug(slug.strip().lower())
    except LookupError as error:
        raise HTTPException(status_code=404, detail="Article not found") from error


@router.post("", response_model=Article, status_code=201)
def post_article(payload: ArticleCreate) -> Article:
    try:
        return create_article(payload)
    except ValueError as error:
        raise HTTPException(status_code=409, detail=str(error)) from error


@router.put("/{article_id}", response_model=Article)
def put_article(article_id: int, payload: ArticleUpdate) -> Article:
    try:
        return update_article(article_id, payload)
    except LookupError as error:
        raise HTTPException(status_code=404, detail="Article not found") from error
    except ValueError as error:
        raise HTTPException(status_code=409, detail=str(error)) from error


@router.patch("/{article_id}/status", response_model=Article)
def patch_article_status(article_id: int, payload: ArticleStatusUpdate) -> Article:
    try:
        return set_article_status(article_id, payload.status)
    except LookupError as error:
        raise HTTPException(status_code=404, detail="Article not found") from error


@router.post("/{article_id}/publish", response_model=Article)
def publish_article(article_id: int) -> Article:
    try:
        return set_article_status(article_id, "published")
    except LookupError as error:
        raise HTTPException(status_code=404, detail="Article not found") from error


@router.post("/{article_id}/draft", response_model=Article)
def draft_article(article_id: int) -> Article:
    try:
        return set_article_status(article_id, "draft")
    except LookupError as error:
        raise HTTPException(status_code=404, detail="Article not found") from error


@router.delete("/{article_id}", status_code=204)
def remove_article(article_id: int) -> None:
    try:
        delete_article(article_id)
    except LookupError as error:
        raise HTTPException(status_code=404, detail="Article not found") from error
